### PR TITLE
Allow EAM to respect PIO_NETCDF_FORMAT

### LINKED ISSUE #2903: CAM does not respect value of PIO_NETCDF_FORMAT XML variable
CAM PIO interface appears to not respect the value of the `PIO_NETCDF_FORMAT` variable in `env_run.xml`. The default for this variable is `64bit_offset`, but valid values include `64bit_data`. When working with very large grids (i.e., ne512) variable sizes become too large to write with the 64bit offset format, and we need to use the 64bit data format (which does not have size limitations). One would hope that setting `PIO_NETCDF_FORMAT=64bit_data` in `env_run.xml` would fix this, but the value appears to not be propagated to `cam_pio_createfile`, because the history, restart, and new initial condition files are all written in the 64bit offset format (confirmed in a ne4 test). I can hard-code `pio_64bit_data` in `cam_pio_utils.F90` to get the desired output type with the following changes:
```
diff --git a/components/cam/src/utils/cam_pio_utils.F90 b/components/cam/src/utils/cam_pio_utils.F90
index a599ac4ef..dc095d03c 100644
--- a/components/cam/src/utils/cam_pio_utils.F90
+++ b/components/cam/src/utils/cam_pio_utils.F90
@@ -1071,7 +1071,7 @@ contains

   subroutine cam_pio_createfile(file, fname, mode_in)
     use pio, only : pio_createfile, file_desc_t, pio_noerr, pio_clobber,      &
-         pio_64bit_offset, pio_iotask_rank
+         pio_64bit_offset, pio_64bit_data, pio_iotask_rank
     use cam_abortutils, only : endrun

     ! Dummy arguments
@@ -1083,10 +1083,11 @@ contains
     integer                                   :: ierr
     integer                                   :: mode

-    mode = ior(PIO_CLOBBER, PIO_64BIT_OFFSET)
-    if (present(mode_in)) then
-      mode = ior(mode, mode_in)
-    end if
+    !mode = ior(PIO_CLOBBER, PIO_64BIT_OFFSET)
+    !if (present(mode_in)) then
+    !  mode = ior(mode, mode_in)
+    !end if
+    mode = ior(PIO_CLOBBER, PIO_64BIT_DATA)
```
and the resulting files have netcdf type `cdf5`. This of course does not deal with the problem of propagating the value of `PIO_NETCDF_FORMAT` down to the CAM PIO interface, but does show that we can get the desired behavior and that there is indeed a problem here. I think this is related to #1544 (writing restart files for ne120 on titan), but the issue I am reporting here is not specific to any specific architecture (my tests were on cori-knl). I'll point out that we needed to hard-code `PIO_64BIT_DATA` for our ne120 MMF simulations on Summit as well. The ability to set this via `env_run.xml` and get the expected behavior would be much appreciated! 64bit data should also probably be the default for large grids.

Simple example to reproduce the bug: configure any F-compset case (ne4 and FC5AV1C-L will work), run `./xmlchange PIO_NETCDF_FORMAT=64bit_data` in the case directory, setup, build, and submit, and then run `ncdump -k` on any of the CAM output files (the restart file will suffice). Note that it will be "64bit offset" as opposed to "cdf5".

### PR DESCRIPTION
**Pull request by brhillman**
Allow EAM to respect the value of the XML variable `PIO_NETCDF_FORMAT` that is set in `env_run.xml`. Previously, setting this variable had no effect on the actual format used to write EAM output files (restart, history, initial conditions), due to calls to `cam_pio_createfile` explicitly passing in a mode argument hard-coded to `PIO_CLOBBER`. Note that there are some whitespace changes in  components/cam/src/dynamics/se/dyn_comp.F90 just because I found some inconsistent indenting when opening that file to remove the hard-coded mode argument in the call to `cam_pio_createfile`. 

Fixes #2903

[BFB]

### CODE REVIEW

**Comment by brhillman** [comment]
This should be BFB, testing now. Will update PR to include BFB label once tests complete.

**Comment by jayeshkrishna** [comment]
You might also want to add someone (@singhbalwinder ?) from the ATM development team as a reviewer

File: components/cam/src/utils/cam_pio_utils.F90 [hunk 1]
```diff
@@ -1089,13 +1091,18 @@ subroutine cam_pio_createfile(file, fname, mode_in)
...
     integer                                   :: mode
     
-    mode = ior(PIO_CLOBBER, PIO_64BIT_OFFSET)
+    ! Get mode to open file in. The default is to get this from shr_pio_mod,
+    ! but an optional mode_in argument is allowed to override this.
     if (present(mode_in)) then
-      mode = ior(mode, mode_in)
+      mode = mode_in
+    else
+      mode = shr_pio_getioformat(atm_id)
```
Comment (jayeshkrishna) [inline components/cam/src/utils/cam_pio_utils.F90]: You might also want to add PIO_CLOBBER by default here

File: components/cam/src/dynamics/se/dyn_comp.F90 [hunk 2]
```diff
@@ -417,40 +417,40 @@ subroutine write_grid_mapping(par, elem)
...
-       ierr = pio_def_var(nc, 'element_corners', PIO_INT, (/dim1,dim2/),vid)
-    
-       ierr = pio_enddef(nc)
-       if (par%dynproc) then
-          call createmetadata(par, elem, subelement_corners)
-       end if
+    ! write meta data for physics on GLL nodes
+    call cam_pio_createfile(nc, 'SEMapping.nc')
 
-       jj=0
```
Comment (jayeshkrishna) [inline components/cam/src/dynamics/se/dyn_comp.F90]: It would be better to move the whitespace changes to a separate commit (easier to view changes on github without "-w")

Comment (brhillman) [reply components/cam/src/dynamics/se/dyn_comp.F90]: Do the last two commits I just pushed fix this? Or do I need to alter the git history somehow to make this play nice?

File: components/cam/src/utils/cam_pio_utils.F90 (hunk 1 above)
Comment (brhillman) [reply components/cam/src/utils/cam_pio_utils.F90]: Maybe with something like `ior(mode, PIO_CLOBBER)` after this logical?

File: components/cam/src/dynamics/se/dyn_comp.F90 (hunk 2 above)
Comment (jayeshkrishna) [reply components/cam/src/dynamics/se/dyn_comp.F90]: Yeah rebasing the branch (two commits, the second commit does not include any white space changes. If you want to include the white space changes add it as a third commit with just the white space changes) would be nicer. Let me know if you need help with it.

File: components/cam/src/utils/cam_pio_utils.F90 (hunk 1 above)
Comment (jayeshkrishna) [reply components/cam/src/utils/cam_pio_utils.F90]: Yes that should work
```
mode = ior(mode, PIO_CLOBBER)
```

File: components/cam/src/dynamics/se/dyn_comp.F90 (hunk 2 above)
Comment (jayeshkrishna) [reply components/cam/src/dynamics/se/dyn_comp.F90]: Thanks the rebased branch looks good

Comment (brhillman) [reply components/cam/src/dynamics/se/dyn_comp.F90]: No problem!

**Comment by brhillman** [comment]
Pinging @singhbalwinder

**Comment by singhbalwinder** [comment]
Hi @brhillman : This looks like BFB change but I was just wondering if you have confirmed it by running a short test.

**Comment by brhillman** [comment]
@singhbalwinder I believe I did but can't find it at the moment...I can re-run to confirm.

**Comment by brhillman** [comment]
@singhbalwinder confirmed this is BFB for a quick smoke test comparison of baselines against master, using `SMS_Ld1_P16x8.ne4_ne4.FC5AV1C-L.cori-knl_intel`:
```
bhillma@cori07:~/codes/e3sm/branches/fix-cam-pioformat/cime/scripts> less  /global/cscratch1/sd/bhillma/acme_scratch/cori-knl/SMS_Ld1_P16x8.ne4_ne4.FC5AV1C-L.cori-knl_intel.C.20190516_153046_ympsi2/TestStatus
PASS SMS_Ld1_P16x8.ne4_ne4.FC5AV1C-L.cori-knl_intel CREATE_NEWCASE
PASS SMS_Ld1_P16x8.ne4_ne4.FC5AV1C-L.cori-knl_intel XML
PASS SMS_Ld1_P16x8.ne4_ne4.FC5AV1C-L.cori-knl_intel SETUP
PASS SMS_Ld1_P16x8.ne4_ne4.FC5AV1C-L.cori-knl_intel SHAREDLIB_BUILD time=301
PASS SMS_Ld1_P16x8.ne4_ne4.FC5AV1C-L.cori-knl_intel NLCOMP
PASS SMS_Ld1_P16x8.ne4_ne4.FC5AV1C-L.cori-knl_intel MODEL_BUILD time=380
PASS SMS_Ld1_P16x8.ne4_ne4.FC5AV1C-L.cori-knl_intel SUBMIT
PASS SMS_Ld1_P16x8.ne4_ne4.FC5AV1C-L.cori-knl_intel RUN time=121
PASS SMS_Ld1_P16x8.ne4_ne4.FC5AV1C-L.cori-knl_intel BASELINE master: 
PASS SMS_Ld1_P16x8.ne4_ne4.FC5AV1C-L.cori-knl_intel TPUTCOMP
PASS SMS_Ld1_P16x8.ne4_ne4.FC5AV1C-L.cori-knl_intel MEMLEAK insuffiencient data for memleak test
PASS SMS_Ld1_P16x8.ne4_ne4.FC5AV1C-L.cori-knl_intel SHORT_TERM_ARCHIVER
```

**Comment by singhbalwinder** [comment]
Thanks @brhillman ! I will merge it to next if it is open today for merging.

**Comment by jayeshkrishna** [comment]
FYI: I just added BFB in the PR description

**Comment by singhbalwinder** [comment]
Merged to next

**Comment by brhillman** [comment]
Thanks @singhbalwinder @jayeshkrishna !
